src/IK/tests/QPInverseKinematicsTest.cpp

```diff
@@ -0,0 +1,193 @@
+
+            for (auto& joint : jointsPos)
+            {
+                joint = iDynTree::getRandomDouble();
+            }
+
+            for (auto& joint : jointsVel)
+            {
+                joint = iDynTree::getRandomDouble();
+            }
+
+            for (auto& element : gravity)
+            {
+                element = iDynTree::getRandomDouble();
+            }
+
+            REQUIRE(kinDyn->setRobotState(worldBasePos, jointsPos, baseVel, jointsVel, gravity));
+
+            // Instantiate the handler
+            VariablesHandler variablesHandler;
+            variablesHandler.addVariable(robotVelocity, model.getNrOfDOFs() + 6);
+
+
+            // Set the frame name
+            const std::string controlledFrame = model.getFrameName(numberOfJoints);
+            parameterHandler->getGroup("SE3_TASK").lock()->setParameter("frame_name", controlledFrame);
+
+            const Eigen::VectorXd kpRegularization = Eigen::VectorXd::Ones(model.getNrOfDOFs());
+            const Eigen::VectorXd weightRegularization = Eigen::VectorXd::Ones(model.getNrOfDOFs());
+            parameterHandler->getGroup("REGULARIZATION_TASK").lock()->setParameter("kp", kpRegularization);
+            parameterHandler->getGroup("REGULARIZATION_TASK").lock()->setParameter("weight", weightRegularization);
+
+            auto ikAndTasks = createIK(parameterHandler, kinDyn, variablesHandler);
+
+            // getCoMPosition and velocity
+            const Eigen::Vector3d desiredCoMPosition = iDynTree::toEigen(kinDyn->getCenterOfMassPosition());
+            const Eigen::Vector3d desiredCoMVelocity = iDynTree::toEigen(kinDyn->getCenterOfMassVelocity());
+
+            const manif::SE3d desiredPose = toManifPose(kinDyn->getWorldTransform(controlledFrame));
+            const manif::SE3d::Tangent desiredVelocity = iDynTree::toEigen(kinDyn->getFrameVel(controlledFrame));
+
+
+            REQUIRE(ikAndTasks.se3Task->setSetPoint(desiredPose, desiredVelocity));
+            REQUIRE(ikAndTasks.comTask->setSetPoint(desiredCoMPosition, desiredCoMVelocity));
+            REQUIRE(ikAndTasks.regularizationTask->setSetpoint(iDynTree::toEigen(jointsPos), iDynTree::toEigen(jointsVel)));
+
+            REQUIRE(ikAndTasks.ik->advance());
+
+            // check that everything went fine
+            REQUIRE(kinDyn->setRobotState(worldBasePos.asHomogeneousTransform(),
+                                          jointsPos,
+                                          ikAndTasks.ik->get().baseVelocity.coeffs(),
+                                          ikAndTasks.ik->get().jointVelocity,
+                                          gravity));
+
+            REQUIRE(ikAndTasks.ik->get().jointVelocity.isApprox(iDynTree::toEigen(jointsVel), tolerance));
+            REQUIRE(desiredCoMVelocity.isApprox(iDynTree::toEigen(kinDyn->getCenterOfMassVelocity()), tolerance));
+            REQUIRE(desiredVelocity.isApprox(iDynTree::toEigen(kinDyn->getFrameVel(controlledFrame)), tolerance));
```

Comment:
To be honest, I am not sure if this is checking if the QPIK is working. You are initializing it already in the solution when setting the robot state in line 149, right?

Reply:
It really depends on what you want to test :wink:

Reply:
As discussed F2F I reimplemented the test in https://github.com/dic-iit/bipedal-locomotion-framework/pull/229/commits/df1e11f3405405211b6a04ccbc9466c297d445f3

Here I run the IK sequentially and I check that the solution converged to the desired one. The tests run in three different models automatically generated by iDynTree.

While I was writing the test I noticed that `osqp-eigen` does not support unconstrained optimization problems. This is actually a limitation since we may want to consider all the tasks in the cost function (If I'm not mistaken the AnDy ik works in this way @lrapetti )

Reply:
Is it a limitation of ``osqp`` or ``osqp-eigen``? It may be worked around with some box constraints on the control variables, eventually with large (but not infinite) bounds.

Reply:
> This is actually a limitation since we may want to consider all the tasks in the cost function (If I'm not mistaken the AnDy ik works in this way @lrapetti )

Yes, I confirm all the task are in the const function. However, if the solver is set to QP we use `osqp-eigen` as well, and we use joint velocity constraints (not sure if you are forced to set them, @kouroshD worked on that part and might know better).

Reply:
> While I was writing the test I noticed that osqp-eigen does not support unconstrained optimization problems. This is actually a limitation since we may want to consider all the tasks in the cost function (If I'm not mistaken the AnDy ik works in this way @lrapetti )


I am not sure why you are making this decision to avoid using constraint in IK. 
Since a constraint is ensured to be respected (with a good sampling time), while a soft constraint added to the task can be violated. Anyhow, as far as I remember and logically, you should be able to solve the optimization problem without constraints as well.

Reply:
That's just a possible implementation. It has pros and cons. For example, you may decide to use no constraints so that you are sure you always get a solution, even if you have two completely contradictory tasks.

Reply:
Note that in a manifold,  if you have a set of constraints such that the possible space of your solution is not empty, it should retrieve a _best solution_ as a result, and it is the task of the designer to do not fix such contradictory constraints!

Reply:
> and it is the task of the designer to do not fix such contradictory constraints!

You should not trust anyone :grin: 

Jokes apart, one thing is what you should do or not, another thing is a limitation introduced by the software. You should not justify a flaw in the design with the excuse that the corresponding use cases are not good enough.

Reply:
https://github.com/robotology/osqp-eigen/issues/84